### DISCUSSION TITLE
Seeking asssitance with base deployment errors using arcgis-enterprise-primary.json file

### DISCUSSION THREAD

**Discussion by J-J-E** [Esri/arcgis-cookbook#387] [Q&A] [answered]
Hello everyone! I'm having a little bit of trouble correctly configuring this base deployment. For some reason, the datastore component fails to configure and says it cant reach the arcgis server component on localhost.  I have validated the url below with the ip:port is indeed accessible, and both the portal and server components are in place and running and appear to be working fine. Im attaching the stack trace as well as the config file (with passwords removed). Any help is appreciated as I've kind of hit a roadblock. Thanks in advance! 

STDERR: Error encountered: Could not connect to server on machine 'localhost'. ArcGIS Data Store or ArcGIS Server on that machine may not be running or the machine is not reachable at this time.
---- End output of "/opt/arcgis/datastore/tools/configuredatastore.sh" "https://10.168.0.23:6443/arcgis/admin" "PSAArcGISServer" "*****" "/opt/arcgis/gisdata/arcgisdatastore" --stores relational ----
Ran "/opt/arcgis/datastore/tools/configuredatastore.sh" "https://10.168.0.23:6443/arcgis/admin" "PSAArcGISServer" "*****" "/opt/arcgis/gisdata/arcgisdatastore" --stores relational returned 1

[cinc-stacktrace.txt](https://github.com/user-attachments/files/15597804/cinc-stacktrace.txt)
[arcgis-enterprise-primary.json](https://github.com/user-attachments/files/15597821/arcgis-enterprise-primary.json)

**Comment by cameronkroeker** [comment]
Hi @J-J-E,

One thing I notice is using IP for server and portal but datastore is set to hostname. In the datastore json block, try changing `"preferredidentifier": "hostname",` to `"preferredidentifier": "ip",`:

```json
"data_store": {
      "install_dir": "/opt",
      "setup_options": "-f Relational,TileCache",
      "data_dir": "/opt/arcgis/gisdata/arcgisdatastore",
      "preferredidentifier": "ip",
      "install_system_requirements": true,
      "types": "relational",
      "tilecache": {
        "backup_type": "none",
        "backup_location": "/net/FILESERVER/gisdata/arcgisbackup/tilecache"
      },
      "relational": {
        "backup_type": "fs",
        "backup_location": "/opt/arcgis/gisdata/arcgisbackup/relational"
      }
    }
```

The other thing is may want to change `FILESERVER` to the actual file server hostname, or ensure there is an `/etc/hosts` entry for it. Or if not wanting to create the backup for the tilecache, set `"backup_type": "fs",` to `"backup_type": "none",`.

Thanks,
Cameron K.

**Reply by J-J-E** [reply]
Cameron, 

Thank you very much for the suggestion. 

1. I updated the identifier from hostname -> ip as suggested. 
2. I removed the tilecache component as we actually have no need for one in this base deployment
2.1 I also removed the reference to the FILERSERVER, as we have no plans to add a standby machine and [per the docs](https://github.com/Esri/arcgis-cookbook/tree/c92c7768637d5e5e2bb633a867107b8e9c206953/templates/arcgis-enterprise-base/11.3#:~:text=If%20you%20don%27t%20plan%20to%20add%20a%20standby%20machine%20in%20the%20future%2C%20don%27t%20configure%20the%20file%20server%2C%20and%20use%20local%20paths%20instead%20of%20shared%20directories%20for%20ArcGIS%20Server%20server%20directories%2C%20Portal%20for%20ArcGIS%20content%20directory%2C%20and%20ArcGIS%20Data%20Store%20backup%20directories%20in%20the%20arcgis%2Denterprise%2Dprimary.json%20file.), local paths should be used instead of a file server in that case
3. I created my own recipe to create local directories, paths are defined in [arcgis][localdirectories][directories]. I updated the enterprise-primary.json file with these paths and the call to the recipe (also attached) as they were in the initial 10.9.1 deployment but I didn't copy them into the 11.3 json file because they already existed. I had to upload the localdirectories.rb file as a .txt file as github wouldn't accept the .rb

This is still failing at the configure portal. The datastore is installed and running, but for some reason cannot reach the ArcGIS Server component on its local host. I have validated I can access the server services and admin pages at the url in the stacktrace that the datastore cannot connect to. 

[arcgis-enterprise-primary1103.json](https://github.com/user-attachments/files/15603525/arcgis-enterprise-primary1103.json)
[localdirectories.txt](https://github.com/user-attachments/files/15603530/localdirectories.txt)
[stacktrace.txt](https://github.com/user-attachments/files/15603533/stacktrace.txt)
[chef-enterprise1103-clioutput.txt](https://github.com/user-attachments/files/15603534/chef-enterprise1103-clioutput.txt)

**Comment by cameronkroeker** [comment]
@J-J-E thanks for the additional logs and information.

One observation is in the localdirectories.rb recipe the following is defined `['arcgis']['localserver']['directories']`, however in the json file its defined as `['arcgis']['localdirectories']['directories']`. So I am not quite sure I understand how that part succeeded.

**localdirectories.rb:**
```ruby
#
# Cookbook Name:: arcgis-enterprise
# Recipe:: createlocaldirs
#
# Copyright 2023 John Ehlen
#

# Create local server directories

node['arcgis']['localserver']['directories'].each do |dir|
  directory dir do
    owner node['arcgis']['run_as_user']
    mode '0755' if node['platform'] != 'windows'
    recursive true
    action :create
  end
end
```


I would expect the json to look like this to match the localdirectories.rb:

```json
"localserver": {
   "directories": [
      "/opt/arcgis/gisdata/arcgisserver",
      "/opt/arcgis/gisdata/arcgisbackup",
      "/opt/arcgis/gisdata/arcgisbackup/relational",
      "/opt/arcgis/gisdata/arcgisportal",
      "/opt/arcgis/gisdata/arcgisportal/content"
   ]
},
```

Alternatively update the localdirectories.rb to match the json:

```ruby
node['arcgis']['localdirectories']['directories'].each do |dir|
  directory dir do
    owner node['arcgis']['run_as_user']
    mode '0755' if node['platform'] != 'windows'
    recursive true
    action :create
  end
end

```

At any rate I suppose it worked somehow and its likely not related to the relational data store failing to register with server. Let's check to ensure both the arcgis server and arcgis data store services are started. As a sudo user run the following two commands on share the result:

- `sudo systemctl status arcgisserver.service`
- `sudo systemctl status arcgisdatastore.service`

Thanks,
Cameron K.

**Reply by J-J-E** [reply]
Cameron, 

the name discrepancy is (obviously) my fault and can be ignored. I originally named it localserver and it made no sense as it wasn't a server, and just local directories, so I renamed it quickly to local directories last night when I uploaded the files and missed a spot lol. My apologies for adding more confusion. I updated it everywhere to say localdirectories. 

Attached are the requested service statuses (along with portal status),  the output of portal and server diag tools. and the updated localdirectories file and enterprise primary json files. 


[arcgisportal-status.txt](https://github.com/user-attachments/files/15666526/arcgisportal-status.txt)
[arcgisdatastore-status.txt](https://github.com/user-attachments/files/15666530/arcgisserver-datastore.txt)
[arcgisserver-status.txt](https://github.com/user-attachments/files/15666533/arcgisserver-status.txt)
[portaldiag.txt](https://github.com/user-attachments/files/15666537/portaldiag.txt)
[serverdiag.txt](https://github.com/user-attachments/files/15666538/serverdiag.txt)
[localdirectories.txt](https://github.com/user-attachments/files/15667208/localdirectories.txt)
[arcgis-enterprise-primary1103.json](https://github.com/user-attachments/files/15667211/arcgis-enterprise-primary1103.json)


UPDATE: 

I also noticed the datastore web page has an insecure self-signed ssl cert, and the page has an error upon loading. 

*RequestError: Unable to load https://10.168.0.23:2443/arcgis/datastoreadmin/configure?f=json&dojo.preventCache=1717696024655 status: 500*


![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/958a20e0-316e-4e63-9dc9-34d9828bca88)


UPDATE 2. 

It also appears that, while the datastore service starts, it wont actually initialize and I cannot replace the ssl cert within it

![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/743fc57e-a900-45e4-a225-52799e95bd6e)

**Reply by cameronkroeker** [reply]
Great find! I think data store installation is corrupted in some way. It may be worth uninstalling and reinstalling. It could be related to perhaps originally the preidentifier being set to hostname instead of ip. Let's try uninstalling/removing datastore and then re-running.

**Reply by J-J-E** [reply]
Uninstalled ArcGIS DS

![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/408fa1a4-04e3-4e38-a3f4-8979843a2fad)

Deleted Leftover folders in DS directory

![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/9cc59eb2-7392-464c-aba1-fbe464e2d699)

Reboot Machine

![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/823a70bb-8baf-4042-ad97-b5e746a29079)

Reran install command routing output to text file

`root@gis-test-enterprise:/opt/cinc# cinc-client -z -j /nfs/InstallMedia/ArcGISEnterprise1103/ChefConfig/arcgis-enterprise-primary1103.json > /nfs/tmp/chef-enterprise1103-clioutput-postDSuninstall.txt`

CLI Output & Stacktrace:

[chef-enterprise1103-clioutput-postDSuninstall.txt](https://github.com/user-attachments/files/15684087/chef-enterprise1103-clioutput-postDSuninstall.txt)
[stacktrace_after_uninstall.txt](https://github.com/user-attachments/files/15684092/stacktrace_after_uninstall.txt)

**Reply by cameronkroeker** [reply]
This is interesting `Could not connect to server on machine 'GIS-TEST.FILMLA.COM'`. I'm not quite sure why it would attempt to connect to that when it should be trying to connect to `10.168.0.23`.

Let's check the following file to ensure it's using ip and isn't set to GIS-TEST.FILMLA.COM.

`/opt/arcgis/datastore/framework/etc/hostidentifier.properties`

In this situation we may also need to define both hostidentifier and preidentifier:
- `node['arcgis']['data_store']['hostidentifier']` = Host name or IP address of ArcGIS Data Store machine.
- `node['arcgis']['data_store']['preferredidentifier']` = ArcGIS Data Store preferred identifier method <hostname|ip>. Default method used is hostname.

```json
"data_store": {
            "install_dir": "/opt",
            "data_dir": "/opt/arcgis/gisdata/arcgisdatastore",
            "preferredidentifier": "ip",
            "hostidentifier": "10.168.0.23",
            "install_system_requirements": true,
            "types": "relational",
            "relational": {
                "backup_type": "fs",
                "backup_location": "/opt/arcgis/gisdata/arcgisbackup/relational"
            }
        }
```

**Reply by J-J-E** [reply]
Existing contents of hostidentifier.properties prior to adding the [hostidentifier] property

![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/871d606b-34d2-4fbd-897a-4cbaa1d327af)

The only thing I can think of is that this is a duplicate/hidden env, in the sense that we have a standing gis-test.filmla.com instance deployed at version 10.9.1 that our dns in cloudflare points to. The original plan was to stand this new instance up side by side, with the same DNS mapping at gis-test-filmla.com (through the hostfile confiuration) to allow me to configure this env, and get it ready, before we cut dns over to point at this machine. Below is the hostfile. I have tried this with the hostfile mapping point gis-test.filmala.com to both the internal and external ip addresses with the same resulting errors. 

![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/6ac01d2d-2434-491a-887f-efc949a86ef5)


CLI Output and Stacktrace with added host ident property


[chef-clioutput-AddHostIdent.txt](https://github.com/user-attachments/files/15685757/chef-clioutput-AddHostIdent.txt)
[stacktrace-AddHostIdent.txt](https://github.com/user-attachments/files/15685758/stacktrace-AddHostIdent.txt)

**Reply by cameronkroeker** [reply]
Thanks. After adding hostidentifier, did it set `hostidentifier=10.168.0.23` in `/opt/arcgis/datastore/framework/etc/hostidentifier.properties`? 

The only other thing I can think of is perhaps when data store reaches out to https://10.168.0.23:6443/arcgis/admin endpoint its getting redirected to https://gis-test.filmla.com/server (since this is server's webcontext url). Is this url accessible from the server/datstore node or only accessible outside? Perhaps /etc/hosts needs to be modified to:

`10.168.0.23 gis-test.filmla.com` instead of `35.235.116.48 gis-test.filmla.com`. Jus a thought Im not sure to be honest.

Update: Just realized you have already tried this! I am not quite sure what the culprit is here. From what I can tell the chef configuration all looks correct and valid.

**Reply by J-J-E** [reply]
1. did it set hostidentifier=10.168.0.23 in /opt/arcgis/datastore/framework/etc/hostidentifier.properties?
  - The ip address was actually inside this file prior to adding the [hostidentifier] in the json file. IT stayed the same after adding that config

2. when data store reaches out to https://10.168.0.23:6443/arcgis/admin endpoint its getting redirected to https://gis-test.filmla.com/server 
 - this redirect isn't happening from my local machine (screenshot below). This is a headless linux vm so I don't have a desktop browser to try accessing it from the machine, but I can't imagine the url isn't accessible as the datastore node is on the same machine as the server node and shares the same IP? I am able to ping the ip address from the cli successfully and I also verified no firewall was active 


![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/e21ae80e-4160-430f-95fd-9979a35d78c1)


![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/c246da11-a4b1-4b47-bd8b-8e2a3709108f)



3. Yes, I have tried configuring with the hostfile pointing to the internal and to the external IP with the same results. I'm kind of at a loss as well here. Been troubleshooting this for two weeks before I posted anything and I've hit a roadblock in this trying to get it configured and running :/

**Reply by cameronkroeker** [reply]
I wonder if it being federated already is causing an issue. Can you unfederate, and then try registering the datastore? 

I imagine manually running the following command yields the same result?

```
/opt/arcgis/datastore/tools/configuredatastore.sh "https://10.168.0.23:6443/arcgis/admin" "PSAArcGISServer" "*****" "/opt/arcgis/gisdata/arcgisdatastore" --stores relational
```

**Reply by cameronkroeker** [reply, marked as answer]
Actually, I just thought of something. Let's check to ensure the hostname.properties file for ArcGIS Server is configured with the correct ip and not set to hostname. I wonder if since originally the /etc/hosts file contained the entry `10.168.0.23 gis-test.filmla.com` when server installed it picked gis-test.filmla.com as its hostname/machine name.

`/opt/arcgis/server/framework/etc/hostname.properties`

Also sign into https://10.168.0.23:6443/arcgis/admin and select machine and see what the machine name is set to.

- node['arcgis']['server']['hostname'] = Host name or IP address of ArcGIS Server machine. Default value is ''.

As well as for portal:
- node['arcgis']['portal']['hostname'] = Host name or IP address of Portal for ArcGIS machine specified in the hostname.properties file. Default value is ''.
- node['arcgis']['portal']['preferredidentifier'] = Portal for ArcGIS preferred identifier method <hostname|ip>. Default method used is hostname.

We may need to start completely over and ensure those attributes are set prior to the installation.

**Reply by J-J-E** [reply]
Cameron, 

Thank you again for all the time you're spending helping me with this. I greatly appreciate it.

1. Same issue running the command manually
2. arcgis server admin page is configured using IP 
![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/1b47dc57-77e2-44ff-96b2-dc0d61b900d5)

3. Arcgis portal appears to be configured using machine fully qualified internal GCP hostname, which is strange, as this isn't in any config file and when I check hostname on the machine, its not the fully qualified hostname but rather just the first part "gis-test-enterprise"
![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/17fd4d3f-8550-462b-844b-539890cb96d4)
![image](https://github.com/Esri/arcgis-cookbook/assets/24575050/321bbcaa-b140-4e90-b15e-c3566f1c4a23)


I think you're probably right in that the best bet is to blow this machine away and redploy a new instance fully from scratch. On that note, can you validate I added the hostname and identifier properties correctly in this updated config? Do I need to add the ssl cert into the [arcgis][server] component? Anything else I should change before I deploy with this script? 

[arcgis-enterprise-primary1103.json](https://github.com/user-attachments/files/15686668/arcgis-enterprise-primary1103.json)

**Reply by cameronkroeker** [reply]
Just need to add a comma to line 43 and 78 of the json file. Other than that is looks good to me. Adding the ssl cert into [arcgis][server] component imports the certificate into arcgis server which is used for the admin url 6443. If you have a cert issued to the ip then I'd say its worth it, otherwise I dont see the need to import a cert that doesn't match https://10.168.0.23:6443/arcgis because then you would get an annoying mismatched address cert error everytime navigating to that url.

**Reply by J-J-E** [reply]
ahh good catch on the commas haha. I did not add the ssl cert to the arcgis server component, ran the install and everything appears to be working correctly with both the server and portal showing the ip address as machine name, and the datastore configured correctly. Thanks again for all your help. I suppose now I'm off to republish 200 services haha!

**Reply by J-J-E** [reply]
@cameronkroeker Hey Cameron, thanks again for your previous help. I am experiencing another issue in this environment configuration and I believe its related to the ssl somehow. I've found [this help article](https://support.esri.com/en-us/knowledge-base/the-option-to-store-credentials-is-not-visible-when-add-000015387#:~:text=Solution%20or%20Workaround%201%20Disable%20the%20firewall%20on,ArcGIS%20to%20trust%20the%20provided%20SSL%20certificates.%20), and Esri Support indicates they believe this is an issue with the SSL yet we cannot find a resolution since everything at the portal endpoints shows a valid cert. They recommended we reconfigure the env using hostnames and not IP addresses, and I was wondering if you could tell me which hostname/FQDNI should use for the URLs, as well as the identifier settings. 

When looking at the machine itself, the hostname is just "gis-test-enterprise", and its fully qualified URL within out LAN in Google Cloud is "https://gis-test-enterprise.us-west2-c.c.ops-gcp.internal".
